src/lib.rs

```diff
@@ -410,4 +461,63 @@ mod tests {
         let ratings = contract.ratings.get(&id).unwrap();
         assert_eq!(ratings.downvote, 1);
     }
+
+    #[test]
+    #[should_panic(expected = "Article not found")]
+    fn donate_to_nonexistent_article() {
+        let context = get_context(vec![], false, ONE_NEAR * 2);
+        testing_env!(context);
+        let mut contract = Wiki::default();
+        contract.donate(0);
+    }
+
+    #[test]
+    #[should_panic(expected = "Donation amount cannot be less than")]
+    fn donate_too_little_to_article() {
+        let context = get_context(vec![], false, ONE_NEAR * 2);
+        testing_env!(context);
+        let mut contract = Wiki::default();
+        let id = contract.create_article(String::from("Title"), String::from("Content"));
+
+        let context = get_context(vec![], false, 999_000_000_000_000_000_000_000); // 0.999 $NEAR
+        testing_env!(context);
+        contract.donate(id);
+    }
+
+    #[test]
+    #[should_panic(expected = "Donation amount cannot be greater than")]
+    fn donate_too_much_to_article() {
+        let context = get_context(vec![], false, ONE_NEAR * 2);
+        testing_env!(context);
+        let mut contract = Wiki::default();
+        let id = contract.create_article(String::from("Title"), String::from("Content"));
+
+        let context = get_context(vec![], false, 100_010_000_000_000_000_000_000_000); // 100.01 $NEAR
+        testing_env!(context);
+        contract.donate(id);
+    }
+
+    #[test]
+    fn donate_successfully_to_article() {
+        let context = get_context(vec![], false, ONE_NEAR * 2);
+        testing_env!(context);
+        let mut contract = Wiki::default();
+        let id = contract.create_article(String::from("Title"), String::from("Content"));
+
+        let context = get_context(vec![], false, ONE_NEAR * 3 / 2);
+        testing_env!(context);
+        contract.donate(id);
+
+        let mut receipts = get_created_receipts();
+        receipts.retain(|r| {
+            let raw_receipt = serde_json::value::to_raw_value(&r).unwrap();
```

Comment:
Why do we need to convert between json?

Reply:
in an ideal world, we could use `serde_json::to_value()` to convert the `receipt` object directly into a JSON object. However, `serde_json` fails to handle `u128` type, so it panics like so:
```
thread 'tests::donate_successfully_to_article' panicked at 'called `Result::unwrap()` on an `Err` value: Error("u128 is not supported", line: 0, column: 0)'
```
Therefore a workaround is currently in place, as you see, where we retrieve the string representation of the `receipt` object, then construct it back into a JSON object, which we can operate further on.